compiler/rustc_span/src/lib.rs

```diff
@@ -1889,22 +1914,35 @@ where
     fn hash_stable(&self, ctx: &mut CTX, hasher: &mut StableHasher) {
         const TAG_VALID_SPAN: u8 = 0;
         const TAG_INVALID_SPAN: u8 = 1;
+        const TAG_RELATIVE_SPAN: u8 = 2;
 
         if !ctx.hash_spans() {
             return;
         }
 
-        self.ctxt().hash_stable(ctx, hasher);
+        let span = self.data();
+        span.ctxt.hash_stable(ctx, hasher);
+        span.parent.hash_stable(ctx, hasher);
 
-        if self.is_dummy() {
+        if span.is_dummy() {
             Hash::hash(&TAG_INVALID_SPAN, hasher);
             return;
         }
 
+        if let Some(parent) = span.parent {
+            let def_span = ctx.def_span(parent).data();
+            if def_span.contains(span) {
+                // This span is enclosed in a definition: only hash the relative position.
+                Hash::hash(&TAG_RELATIVE_SPAN, hasher);
```

Comment:
If I understand correctly, any queries that have a `Span` in their input will *not* get re-run if only the parent definition is moved around. However, the exact line number can show up in diagnostic outputs, so this will cause us to output stale diagnostic messages.

Reply:
Actually, diagnostics hold a `Span`, so their position is automatically updated. To the contrary, everything that touches the source map (including snippets inside the diagnostic) need to be recomputed.